src/vs/vscode.proposed.d.ts

```diff
@@ -33,7 +33,37 @@ declare module 'vscode' {
 		extensionHostEnv?: { [key: string]: string | null };
 	}
 
-	export type ResolverResult = ResolvedAuthority & ResolvedOptions;
+	export interface Port extends Disposable {
+		remotePort: number;
+		localPort?: number;
+		description?: string;
+	}
+
+	/**
+	 * Used as part of the ResolverResult if the extension has any candidate, published, or forwarded ports.
+	 */
+	export interface PortInformation {
+		/**
+		 * Ports the extension knows are interesting and might be nice to forward.
+		 */
+		candidates?: Port[];
+		/**
+		 * Ports that are already immutably forwarded.
+		 */
+		published?: Port[];
+		/**
+		 * An array of ports to be forwarded once connected. For example,
+		 * if the extension wants to have a configuration file with some default ports, they could be passed through for forwarding here.
+		 */
+		toForward?: { remote: number, local: number, name: string }[];
+		/**
+		 * Allows ports that have been forwarded in a workspace to be restored next time that workspace is opened.
+		 * Alternative to this: have API for getting all the ports then an event when they change so that the extension can save and restore ports(as it can do today).
+		 */
+		restorePorts: boolean;
```

Comment:
Need to distinguish between the vscode server port which shouldn't be restored, and others. The alternative suggestion might be better so the extension can differentiate.

Reply:
From what I understand, the vscode server port _must_ be forwarded by the extension. This API deliberately doesn't provide a way to tell the core about that port, so it won't know about it to restore it.

What other ports shouldn't be restored?

Reply:
IMHO `restorePorts: boolean` should be a user setting, not something a provider gets to decide.

Reply:
We were talking about this with Sri today, that if we switch to using this feature, how would you support the ports that already been shared in a session but because user closed the vso (hence vscode), the ports have been killed. so I assume this restorePorts would solve this issue, am I right? but then where do you save the forwarded port to forward them for next session?

Reply:
@fara-nak Instead of having `restorePorts` in the API, I'd like to make it a setting. So when the user opens a workspace again the ports are automatically re-forwarded.